File: chat-web/src/lib/relative-time.ts
/** Sidebar/tab timestamps. Korean, deliberately coarse -- this is a glanceable
 *  hint, not a log. `now` is injected so it is testable without faking clocks. */
export function relativeTime(epochMs: number | null, now: number): string {
  if (epochMs === null) return "";
  const diff = now - epochMs;
  if (diff < 0) return "방금";
  const min = Math.floor(diff / 60_000);
  if (min < 1) return "방금";
  if (min < 60) return `${min}분`;
  const hour = Math.floor(min / 60);
  if (hour < 24) return `${hour}시간`;
  const day = Math.floor(hour / 24);
  if (day === 1) return "어제";
  if (day < 7) return `${day}일`;
  const d = new Date(epochMs);
  return `${d.getMonth() + 1}월 ${d.getDate()}일`;
}

File: pet-app/Puck/ClientWindow/Chat/RelativeTime.swift
//
//  RelativeTime.swift
//  Puck
//
//  Sidebar/tab timestamps. Port of chat-web's relative-time.ts, brought over
//  with the chat UI (2026-08-15) rather than reimplemented -- its rules are
//  the ones the sidebar was designed around.
//
//  Deliberately coarser than RelativeDateTimeFormatter: this is a glanceable
//  hint next to a session title, so "12분" beats "12 minutes ago", and
//  anything older than a week collapses to a date.
//
//  `now` is injected so it is testable without faking the clock -- the same
//  reason the TS version took it as a parameter.
//

import Foundation

enum RelativeTime {
    static func short(since date: Date?, now: Date = Date()) -> String {
        guard let date else { return "" }
        let seconds = now.timeIntervalSince(date)
        // A clock that moved backwards (NTP correction, timezone edit) must
        // not render a negative age.
        guard seconds >= 0 else { return "방금" }

        let minutes = Int(seconds / 60)
        if minutes < 1 { return "방금" }
        if minutes < 60 { return "\(minutes)분" }

        let hours = minutes / 60
        if hours < 24 { return "\(hours)시간" }

        let days = hours / 24
        if days == 1 { return "어제" }
        if days < 7 { return "\(days)일" }

        let components = Calendar.current.dateComponents([.month, .day], from: date)
        guard let month = components.month, let day = components.day else { return "" }
        return "\(month)월 \(day)일"
    }
}
